dhall/src/syntax/ast/visitor.rs

```diff
@@ -31,7 +31,7 @@ where
     // Can't use closures because of borrowing rules
     macro_rules! expr {
         () => {
-            |e| Ok(expr!(e))
+            |e| f(None, e)
```

Comment:
These changes were prompted by more clippy warnings. Even though each usage of the `?` operator made sense in context of the macro, once expanded the linter only considers the expanded expression.

```
warning: Question mark operator is useless here
  --> dhall/src/syntax/ast/visitor.rs:48:25
   |
48 |             opt($e, |e| Ok(expr!(e)))?
   |                         ^^^^^^^^^^^^ help: try: `f(None, $e)`
...
57 |         Let(l, t, a, e) => Let(l.clone(), opt!(t), expr!(a), expr!(l, e)),
   |                                           ------- in this macro invocation
   |
   = note: `#[warn(clippy::needless_question_mark)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_question_mark
   = note: this warning originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

I found it somewhat difficult to unpick, so I ended up inlining the definitions, and finding the smallest subsert of changes that avoided said warnings.

Reply:
For this I'd rather silence the warning, I think the macros really help legibility. Frankly I4d silence the warning for the whole file at this point. I liberally tell clippy to shut up when it's not clearly an improvement

Reply:
Cool–I've just added an `#[allow(…)]` attribute to that function. Hopefully that works?

Reply:
Could you add it at the top of `dhall/src/lib.rs` instead? I don't find this lint helpful in general

Reply:
And then reverse your changes to add the macros back plz

Reply:
Ah looks like you did, but there's one you missed

Reply:
Ah, thanks. I'd missed the one in `dhall/src/operations/kind.rs`.